cpp/src/arrow/ipc/adapter.cc

```diff
@@ -63,32 +65,56 @@ static bool IsPrimitive(const DataType* type) {
   }
 }
 
+static bool IsListType(const DataType* type) {
+  DCHECK(type != nullptr);
+  switch (type->type) {
+    // TODO(emkornfield) grouping like this are used in a few places in the
+    // code consider using pattern like:
+    // http://stackoverflow.com/questions/26784685/c-macro-for-calling-function-based-on-enum-type
+    //
+    // TODO(emkornfield) Fix type systems so these are all considered lists and
+    // the types behave the same way?
+    // case Type::BINARY:
+    // case Type::CHAR:
+    case Type::LIST:
```

Comment:
See also @jacques-n comment on the mailing list. Having logical types for String (UTF8) and Binary makes sense, but internally these are represented in memory using the same memory layout as List<UInt8>. So how we implement them in C++ should be whatever is the most convenient and high performance thing, probably